libsel4vmmplatsupport/src/drivers/virtio_blk.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2020, Data61
+ * Commonwealth Scientific and Industrial Research Organisation (CSIRO)
+ * ABN 41 687 119 230.
+ *
+ * This software may be distributed and modified according to the terms of
+ * the BSD 2-Clause license. Note that NO WARRANTY is provided.
+ * See "LICENSE_BSD2.txt" for details.
+ *
+ * @TAG(DATA61_BSD)
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <platsupport/io.h>
+
+#include <sel4vmmplatsupport/drivers/virtio.h>
+#include <sel4vmmplatsupport/drivers/virtio_blk.h>
+
+#include <pci/helper.h>
+#include <sel4vmmplatsupport/drivers/pci_helper.h>
+
+#define QUEUE_SIZE 128
+
+static ps_io_ops_t ops;
+
+static int virtio_blk_io_in(void *cookie, unsigned int port_no, unsigned int size, unsigned int *result)
+{
+    virtio_blk_t *blk = (virtio_blk_t *)cookie;
+    unsigned int offset = port_no - blk->iobase;
+    unsigned int val;
+    int err = blk->emul->io_in(blk->emul, offset, size, &val);
+    if (err) {
+        return err;
+    }
+    *result = val;
+    return 0;
+}
+
+static int virtio_blk_io_out(void *cookie, unsigned int port_no, unsigned int size, unsigned int value)
+{
+    int ret;
+    virtio_blk_t *blk = (virtio_blk_t *)cookie;
+    unsigned int offset = port_no - blk->iobase;
+    ret = blk->emul->io_out(blk->emul, offset, size, value);
+    return ret;
+}
+
+static int emul_driver_init(struct disk_driver *driver, ps_io_ops_t io_ops, void *config)
+{
+    virtio_blk_t *blk = (virtio_blk_t *)config;
+    driver->disk_data = config;
+    driver->dma_alignment = sizeof(uintptr_t);
+    driver->i_fn = blk->emul_driver_funcs;
+    blk->emul_driver = driver;
+    return 0;
+}
+
+static void *malloc_dma_alloc(void *cookie, size_t size, int align, int cached, ps_mem_flags_t flags)
+{
+    assert(cached);
+    int error;
+    void *ret;
+    error = posix_memalign(&ret, align, size);
+    if (error) {
+        return NULL;
+    }
+    return ret;
+}
+
+static void malloc_dma_free(void *cookie, void *addr, size_t size)
+{
+    free(addr);
+}
+
+static uintptr_t malloc_dma_pin(void *cookie, void *addr, size_t size)
+{
```

Comment:
ZFLOGF

Reply:
This probably shouldn't have ZF_LOGF.

Reply:
What do we normally do when the pin doesn't do anything?

Reply:
It sort of is doing something in a sense, pinning memory is supposed to be mapping into a place of memory where it can't be touched until that particular DMA transaction has finished. In this case, pinning malloc memory doesn't require any extra operation since we don't expect to unmap the malloc'd memory.